File: android/src/main/kotlin/com/courier/courier_flutter/NotificationConverter.kt
package com.courier.courier_flutter

import com.google.firebase.messaging.RemoteMessage

val RemoteMessage.pushNotification: Map<String, Any?>
    get() {

        val rawData = data.toMutableMap()
        val payload = mutableMapOf<String, Any?>()

        // Add existing values to base map
        // then remove the unneeded keys
        val baseKeys = listOf("title", "subtitle", "body", "badge", "sound")
        baseKeys.forEach { key ->
            payload[key] = data[key]
            rawData.remove(key)
        }

        // Add extras
        for ((key, value) in rawData) {
            payload[key] = value
        }

        // Add the raw data
        payload["raw"] = data

        return payload

    }

File: ios/Classes/NotificationConverter.swift
//
//  MessageReducer.swift
//  courier_flutter
//
//  Created by Michael Miller on 10/13/22.
//

import Foundation

extension UNNotificationContent {
    
    var pushNotification: [AnyHashable : Any?] {
        get {
            
            // Initial payload
            var payload: [AnyHashable : Any?] = [
                "title": title,
                "subtitle": nil,
                "body": body,
                "badge": badge,
                "sound": nil
            ]
            
            // Do not add subtitle if it's empty
            if (!subtitle.isEmpty) {
                payload["subtitle"] = subtitle
            }
            
            // Add sound as a string
            if let aps = userInfo["aps"] as? [AnyHashable : Any?], let sound = aps["sound"] {
                payload["sound"] = sound
            }
            
            // Merge the payload data
            // This appends all custom attributes
            var data = userInfo
            data.removeValue(forKey: "aps")
            data.forEach { payload[$0] = $1 }
            
            // Add the raw data
            payload["raw"] = userInfo
            
            return payload
            
        }
    }
    
}
